vncviewer/touch.cxx

```diff
@@ -0,0 +1,324 @@
+/* Copyright 2019 Pierre Ossman <ossman@cendio.se> for Cendio AB
+ * Copyright 2019 Aaron Sowry for Cendio AB
+ * 
+ * This is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ * 
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ * 
+ * You should have received a copy of the GNU General Public License
+ * along with this software; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307,
+ * USA.
+ */
+
+#ifdef HAVE_CONFIG_H
+#include <config.h>
+#endif
+
+#include <assert.h>
+
+#include <map>
+
+#if defined(WIN32)
+#include <windows.h>
+#include <commctrl.h>
+#elif !defined(WIN32) && !defined(__APPLE__)
+#include <X11/extensions/XInput2.h>
+#include <X11/extensions/XI2.h>
+#endif
+
+#include <FL/Fl.H>
+#include <FL/x.H>
+
+#include <rfb/LogWriter.h>
+
+#include "i18n.h"
+#include "vncviewer.h"
+#include "BaseTouchHandler.h"
+#if defined(WIN32)
+#include "win32TouchHandler.h"
+#elif defined(__APPLE__)
+#else
+#include "XInputTouchHandler.h"
+#endif
+
+#include "touch.h"
+
+static rfb::LogWriter vlog("Touch");
+
+#if !defined(WIN32) && !defined(__APPLE__)
+static int xi_major;
+#endif
+
+typedef std::map<Window, class BaseTouchHandler*> HandlerMap;
+static HandlerMap handlers;
+
+#if defined(WIN32)
+LRESULT CALLBACK win32_gestureProc(HWND hWnd, UINT uMsg, WPARAM wParam,
```

Comment:
Can't this be in the Win32 class?

Reply:
Most of it could be moved but not everything. We need to keep a subclass proc function in touch.cxx since we need some way to keep track of the handlers for each window.

The gesture related stuff was moved in 790fb8e6dbd4575a2892f3411e0593ce26ca30e5

If we were to move everything to the Win32 class we would not be able to clean up handlers when a window was closed. The only way to track if a window is closing is to hook into WndProc. The WM_DESTROY and WM_NCDESTROY messages are sent directly to WndProc.